_posts/2019/2019-03-28-be-aware-of-nil-values-when-using-where-not.md

```diff
@@ -0,0 +1,45 @@
+---
+title: Be Aware of nil values when using where.not()
+description: Strange Think
+layout: article
+category: ruby
+image:
+  base: '2019/be-aware-of-nil-values-when-using-where-not'
+  alt: 'Half-finished painted wall'
+  source: "https://unsplash.com/photos/oBb-Y26PJgg"
+  credit: "Samuel Zeller"
+---
+
+The use of `where.not` when building Active Record scopes can help you to elegantly specify scopes where an attribute does not have a specific value.
+
+However the abstraction has some quirks if you have fields that can be `NULL` in your database (`nil` in Ruby).
+
+
+## Instead of…
+
+…expecting `where.not` to return `nil` values:
+
+```ruby
+non_soy_options = Coffee.where.not(milk: "Soy")
+```
+
+
+## Use…
+
+…an `or` clause to explicitly request the `nil` values.
+
+```ruby
+non_soy_options = Coffee.where.not(milk: "Soy").or(Coffee.where(milk: nil))
+```
+
+
+## But why?
+
+Of course this all depends on whether the answer you're looking for is expected to include nil values!
+
+This is because in SQL the results of `WHERE milk = "Soy"` aren’t the inverse of `WHERE milk != "Soy"`. Neither of them include the case `WHERE milk IS NULL`
```

Comment:
But obviously `WHERE milk = "Soy"` wouldn't include cases where milk is NULL...so the 'neither of them' statement doesn't make sense to me.

And to confirm: `WHERE milk != "Soy"` will also not include cases where milk is NULL, right?